### PR TITLE
Add reduction support for complex datatypes

### PR DESCRIPTION
**Pull request by nusislam**
## Pull Request Description

This PR adds support for reduction operations (SUM, PROD, REPLACE) for complex types.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

**Comment by nusislam** [comment]
This PR depends on PR #177. So #177 should be merged before this one.

**Comment by nusislam** [comment]
The tests with `complex` type failed because the result validation in the existing `pack` test is not correct for `complex` PROD operation. PR #177 has a fix for that.

File: src/backend/gencomm.py [hunk 1]
```diff
@@ -93,7 +96,14 @@ def switcher_builtin(backend, OUTFILE, blklens, builtin_types, builtin_maps, typ
...
     for b in builtin_types:
-        switcher_builtin_element(backend, OUTFILE, blklens, typelist, pupstr, "YAKSA_TYPE__%s" % b.replace(" ", "_"), b.replace(" ", "_"))
+        if (b == "c_complex"):
+            switcher_builtin_element(backend, OUTFILE, blklens, typelist, pupstr, "YAKSA_TYPE__C_COMPLEX", b.replace(" ", "_"))
+        elif (b == "c_double_complex"):
+            switcher_builtin_element(backend, OUTFILE, blklens, typelist, pupstr, "YAKSA_TYPE__C_DOUBLE_COMPLEX", b.replace(" ", "_"))
+        elif (b == "c_long_double_complex"):
+            switcher_builtin_element(backend, OUTFILE, blklens, typelist, pupstr, "YAKSA_TYPE__C_LONG_DOUBLE_COMPLEX", b.replace(" ", "_"))
+        else:
+            switcher_builtin_element(backend, OUTFILE, blklens, typelist, pupstr, "YAKSA_TYPE__%s" % b.replace(" ", "_"), b.replace(" ", "_"))
```
Comment (hzhou) [inline src/backend/gencomm.py:106]: Why do we need separate cases? It appears the `else` branch will work for the complex names, right?

File: src/backend/seq/genpup.py [hunk 2]
```diff
@@ -122,6 +122,7 @@ def generate_kernels(b, darray, blklen):
 
     # individual blocklength optimization is only for
     # hvector and blkhindx
+
```
Comment (hzhou) [inline src/backend/seq/genpup.py:125]: Extra white space

File: test/pack/pack.c [hunk 3]
```diff
@@ -390,7 +395,11 @@ void *runtest(void *arg)
                 break;
 
             case YAKSA_OP__PROD:
-                rc = DTP_obj_buf_check(dobj, dbuf_h, 0, -1, basecount);
+                if (strstr(typestr, "complex") != NULL) {
+                    rc = DTP_obj_buf_check(dobj, dbuf_h, 0, 0, basecount);
+                } else {
+                    rc = DTP_obj_buf_check(dobj, dbuf_h, 0, -1, basecount);
```
Comment (hzhou) [inline test/pack/pack.c:401]: Just for my education, can you explain how the buffer is computed?

**Comment by nusislam** [comment]
@hzhou - Thanks for the reviews. I will take a look and get back to you shortly.

File: src/backend/gencomm.py (hunk 1 above)
Comment (nusislam) [reply src/backend/gencomm.py:106]: Agreed. I will fix it.

File: test/pack/pack.c (hunk 3 above)
Comment (nusislam) [reply test/pack/pack.c:401]: I cherry-picked this one from PR #177. @zhenggb72 - Can you please explain?

Comment (zhenggb72) [reply test/pack/pack.c:401]: Originally the two buffers are initialized with 0, 1, 2, 3 ...  and  -1, -1 ..... And the result is  0, -1, -2, -3, -4 ...

For complex case, each complex takes two numbers  real part and  img part. so the result of product is not as simple as 0, -1, -2. So it is wrong.
This commit tried to fix it in a very simple way, just to make it work.

**Comment by nusislam** [comment]
@hzhou - I have addressed your comments.

File: test/pack/pack.c (hunk 3 above)
Comment (hzhou) [reply test/pack/pack.c:401]: I see. Could you add a comment at line 226 -- "Multiply 0 since there is no easy way to verify complex multiplication"?

**Review by hzhou — CHANGES_REQUESTED** [review]
One last comment. Could you also fix the `DCO` check -- just need add the sign-off to the commit messages?

**Comment by nusislam** [comment]
The `yaksa-warning` test failed with the following error:
```fatal error: error writing to /tmp/cccMkLqd.s: No space left on device```

**Review by hzhou — APPROVED** [review]
Alright, I think it is good enough.

**Comment by hzhou** [comment]
test:yaksa/warnings

**Comment by zhenggb72** [comment]
I thought this PR is supposed to be merged after #177

**Comment by hzhou** [comment]
> I thought this PR is supposed to be merged after #177

Why?

**Comment by zhenggb72** [comment]
178 cherry pick from 177

**Comment by hzhou** [comment]
> 
> 
> 178 cherry pick from 177

You referring to the dtp commit? That commit fixes complex testing, it belongs here, right?

**Comment by zhenggb72** [comment]
I saw a conflict now in 177, I will take a closer look.

**Pull request merged**